### PR TITLE
Fix: Expose error messages in two screens Connect and Overview 

### LINKED ISSUE #2266: No error message shown when initiating Stripe redirect from non-Settings page
If "Edit details" redirect link on "Overview" page fails, no error message will show, because errors are only handled by specifically adding an error message to the settings page:

https://github.com/Automattic/woocommerce-payments/blob/add641f73da3863d17d0b831ea35054e8b7c4f9f/includes/class-wc-payments-account.php#L320-L327

This may or may not also be an issue on the Connect screen with the "Finish Setup" button, which uses the same `add_notice_to_settings_page` function.

Marking low priority, partly because the error isn't terribly useful anyway.

### PR DESCRIPTION
**Pull request by htdat**
Fixes #2266

#### Changes proposed in this Pull Request

Stop outputting notices to WcAdmin-powered screens (Connect and Overview) by this [method `add_notice_to_settings_page`](https://github.com/Automattic/woocommerce-payments/blob/33b010ef8c898b3bef12a62d55c8db9a5f089b6e/includes/class-wc-payments-account.php#L749-L757) as it will be hidden by WcAdmin - [details here](https://github.com/Automattic/woocommerce-payments/issues/2266#issuecomment-880552944). 

Then expose these errors to those WcAdmin-powered screens:

- For Payments > Overview screen - when clicking **Edit Details**, if there is an error in retrieving the Stripe Connect URL, we redirect it to the Overview page with an additional parameter `wcpay-login-error=1`. The React code will check this parameter and display the respective message. 
- For Payments > Connect screen - that's pretty straightforward, switching from `add_notice_to_settings_page` to `redirect_to_onboarding_page` is enough.

#### Testing instructions

Apply this PR first then follow up each test case. 

**Case 1: Edit Details in page Overview**

- Add `throw new Exception('just_a_test');` right above this line to replicate the error:
https://github.com/Automattic/woocommerce-payments/blob/add641f73da3863d17d0b831ea35054e8b7c4f9f/includes/class-wc-payments-account.php#L321
- Go to Payments > Overview > Click `Edit Details`.
- Expectation: see the error message `There was a problem redirecting you to the account dashboard. Please try again.`.

Markup on 2021-07-20 at 18:06:23

**Case 2: Init Stripe oauth flow failed in screen Connect** 

- Add `throw new Exception('just_a_test');` right above this line to replicate the error:
https://github.com/Automattic/woocommerce-payments/blob/add641f73da3863d17d0b831ea35054e8b7c4f9f/includes/class-wc-payments-account.php#L355
- In `WCPay Dev`, choose this option **Force the plugin to act as disconnected from WCPay**. 
- Go to Payments, will be forced to screen Connect.
- Click `Finish Setup`.
- Expectation: see this error: 

Markup on 2021-07-20 at 18:11:54

**Case 3: finalize_connection failure in screen Connect** 

- In `WCPay Dev`, choose this option **Force the plugin to act as disconnected from WCPay**. 
- Visit this page wp-admin/admin.php?page=wc-admin&path=%2Fpayments%2Fconnect&wcpay-state=1&wcpay-mode=1 . In reality, since this is an API redirect back after the OAuth Stripe flow, there are more parameters but this is enough to demonstrate the error.
- Expectation: see this error: 

Markup on 2021-07-20 at 18:15:39

*

-------------------

- [x] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

### CODE REVIEW

**Review by kalessil — APPROVED** [review]
Looks great, thanks for fixing it! 
Preapproving: please address my comment about changelog entries before merging.

File: includes/class-wc-payments-account.php [hunk 1]
```diff
@@ -320,6 +320,18 @@ public function maybe_handle_oauth() {
...
 			} catch ( Exception $e ) {
+				// Handle the error within the WooCommerce Admin-powered app.
+				if ( self::is_overview_page() ) {
+					wp_safe_redirect(
+						add_query_arg(
+							[ 'wcpay-login-error' => '1' ],
+							self::get_overview_page_url()
+						)
+					);
+					exit;
```
Comment (naman03malhotra) [inline includes/class-wc-payments-account.php:331]: As discussed on the call, if we can include a test for this change.

File: includes/class-wc-payments-account.php [hunk 2]
```diff
@@ -320,6 +320,18 @@ public function maybe_handle_oauth() {
 			try {
 				$this->redirect_to_login();
 			} catch ( Exception $e ) {
+				// Handle the error within the WooCommerce Admin-powered app.
```
Comment (naman03malhotra) [inline includes/class-wc-payments-account.php:323]: Nit:
```suggestion
				// Handle the error within the React powered app.
```

File: includes/class-wc-payments-account.php [hunk 3]
```diff
@@ -320,6 +320,18 @@ public function maybe_handle_oauth() {
...
+				if ( self::is_overview_page() ) {
+					wp_safe_redirect(
```
Comment (naman03malhotra) [inline includes/class-wc-payments-account.php:325]: For future reference one other way would have been:

https://github.com/Automattic/woocommerce-payments/blob/aabf276642607ea26491dac05babadf29927b63a/includes/class-wc-payments-account.php#L238

https://github.com/Automattic/woocommerce-payments/blob/aabf276642607ea26491dac05babadf29927b63a/includes/admin/class-wc-payments-admin.php#L253

**Comment by naman03malhotra** [comment]
LGTM, works well!

File: changelog.txt [hunk 4]
```diff
@@ -3,6 +3,7 @@
 = 2.8.0 - 2021-xx-xx =
 * Add - Use date picker for applicable dispute evidence fields.
 * Fix - Disabled Payment Request Buttons when order has to be split into multiple packages because Payment Requests do not support that use case.
+* Fix - Expose error messages in two screens Connect and Overview.
```
[the remark this answers is not in the document]
Comment (htdat) [reply changelog.txt]: 👍 handled in 2732bd1

File: includes/class-wc-payments-account.php (hunk 2 above)
Comment (htdat) [reply includes/class-wc-payments-account.php]: 👍

**Comment by htdat** [comment]
Addressed all comments. 
Can you have a look at my test for handling param `wcpay-login-error=1`?

**Comment by naman03malhotra** [comment]
>Can you have a look at my test for handling param wcpay-login-error=1?

The test is written well 🥳, do mind writing one more for the success notice?

**Review by naman03malhotra — APPROVED** [review]

**Comment by htdat** [comment]
> The test is written well 🥳, do mind writing one more for the success notice?

👍 Addressed here ebb0fe6

**Comment by kalessil** [comment]
Updating the labels here to "ready to merge"

**Pull request merged**